### ISSUE TITLE
server-cuda-b9014 on NVIDIA GB10 (Blackwell, sm_120) aarch64: qwen3-coder-next-Q4_K_M.gguf emits all-`?` tokens (other MoE Q4_K_M GGUFs OK)

### ISSUE THREAD

**Issue by Lightheartdevs** [ggml-org/llama.cpp#23010]
## Symptom

On NVIDIA DGX Spark (GB10 Grace Blackwell, compute capability 12.1, aarch64), `llama-server` from the `ghcr.io/ggml-org/llama.cpp:server-cuda-b9014` image loads `qwen3-coder-next-Q4_K_M.gguf` successfully, accepts chat completions, but every output token is `?` (Unicode U+003F). SHA256 of the GGUF matches upstream. Prompt processing log timings are normal (~220 tok/s prefill). Decode throughput is normal (~63 tok/s). The model just produces `?` forever.

```
$ curl -sS -X POST http://127.0.0.1:8080/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"qwen3-coder-next-Q4_K_M.gguf",
       "messages":[{"role":"user","content":"What is 2+2?"}],
       "max_tokens":40,"temperature":0.2}' \
  | jq '.choices[0].message.content'
"????????????????????????????????????????"
```

System fingerprint: `b9014-d4b0c22f9`. 40 completion tokens, all `?`.

## Counter-test: another MoE Q4_K_M on the same setup works

To rule out a general "MoE-Q4_K_M on Blackwell-aarch64" kernel bug, I served `Qwen3.6-35B-A3B-UD-Q4_K_M.gguf` (unsloth, SHA `ac0e2c1189e055faa36eff361580e79c5bd6f8e76bffb4ce547f167d53e31a61`) on the **same image, same GPU, same flags**:

```
$ curl ... 'What is 2+2?' ...
"2+2 equals 4."
```

Coherent output, 7 tok/s cold / 59 tok/s warm decode across multiple prompts. So the kernel path itself works on this GPU+arch for MoE Q4_K_M; the failure is specific to the `qwen3-coder-next-Q4_K_M.gguf` build.

## Other-platform cross-check

The same `qwen3-coder-next-Q4_K_M.gguf` runs cleanly on RTX 5090 × 2 (Blackwell, sm_120, **amd64**) at ~74 tok/s decode using the same `server-cuda-b9014` image family. So the issue is the **aarch64 × coder-next** intersection, not a sm_120 problem in general.

## Environment

| Item | Value |
|------|-------|
| Hardware | NVIDIA DGX Spark (GB10 Grace Blackwell) |
| GPU compute capability | sm_120 (12.1) |
| NVIDIA driver | 580.142 |
| Host kernel | Linux 6.17.0-1014-nvidia aarch64 |
| Host OS | Ubuntu 24.04 |
| Container image | `ghcr.io/ggml-org/llama.cpp:server-cuda-b9014` |
| `llama-server --version` | `9014 (d4b0c22f9)` · `built with GNU 14.2.0 for Linux aarch64` |
| Backends loaded | `libggml-cuda.so`, `libggml-cpu-armv8.6_2.so` |

## Reproduction

```
docker run --rm --gpus all -p 127.0.0.1:8080:8080 \
  -v /path/to/models:/models:ro \
  ghcr.io/ggml-org/llama.cpp:server-cuda-b9014 \
  --model /models/qwen3-coder-next-Q4_K_M.gguf \
  --host 0.0.0.0 --port 8080 --n-gpu-layers 999 \
  --ctx-size 131072 --batch-size 2048 --threads 4 --parallel 1
```

Then hit `/v1/chat/completions` with any prompt → all `?` tokens.

Server logs show no errors during load or generation. `model loaded`, `server is listening`, `all slots are idle` after each completion.

## What I'm guessing

Some interaction between the coder-next quantisation pattern (Q4_K_M with this particular tensor shape mix) and the CUDA kernels in the aarch64 build of `b9014`. Possibly a wrong kernel selected at runtime for one of coder-next's MoE tensor shapes that the A3B file happens not to hit. I haven't traced further.

Happy to test patches, run different builds, dump model metadata — let me know what would help narrow it down.

**Comment by tbraun96** [comment]
Disclosure: I work on [Atlas](https://github.com/Avarok-Cybersecurity/atlas) and we serve qwen3-coder-next on GB10, so the all-`?` output is a failure mode I recognize. The "loads fine, accepts requests, every single token is U+003F" pattern is almost always a silently miscomputing GPU kernel, not a tokenizer or model-file problem.

The mechanism: the `server-cuda-b9014` image is built for `sm_120`, and GB10 reports `sm_121` (compute 12.1). Rather than failing with "no kernel image", a close-enough cubin can load and run but compute numerically wrong results, the logits come out as NaN or garbage, and the sampler then collapses to a fallback/unknown token that detokenizes to `?`. So the `?` is not a decode bug, it is the visible end of upstream garbage logits. We have seen the exact same sm_120-artifact-on-sm_121 silent miscompute bite coder-next specifically.

The clean way to confirm it is a GPU-kernel issue and not the GGUF:

```bash
# force CPU-only generation; if tokens come out correct, the GGUF is fine and the
# GPU sm_120-vs-sm_121 kernel path is the culprit
llama-server -m qwen3-coder-next-Q4_K_M.gguf -ngl 0
# also worth: a build that actually targets sm_121 (CMAKE_CUDA_ARCHITECTURES=121)
# rather than the prebuilt sm_120 image
```

@Lightheartdevs does `-ngl 0` (CPU-only) produce correct tokens for you? If CPU is clean and GPU is all-`?`, this is the sm_121 kernel-image mismatch and the fix is a build targeting compute 12.1 rather than the stock sm_120 container. If CPU is *also* `?`, then it is the GGUF/tokenizer and we look somewhere completely different.
